aggregators/writer.go

```diff
@@ -0,0 +1,185 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License 2.0;
+// you may not use this file except in compliance with the Elastic License 2.0.
+
+// Package aggregators holds the logic for doing the actual aggregation.
+package aggregators
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/cockroachdb/pebble"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric"
+
+	"github.com/elastic/apm-aggregation/aggregationpb"
+	"github.com/elastic/apm-data/model/modelpb"
+)
+
+// ErrWriterClosed is returned by Writer methods after the writer is closed.
+var ErrWriterClosed = errors.New("writer is closed")
+
+// Writer provides methods for writing metrics to Pebble.
+type Writer struct {
+	aggregator *Aggregator
+
+	mu    sync.Mutex
+	batch *pebble.Batch
+}
+
+// Close commits and closes the writer's batch,
+// and removes the writer from the aggregator.
+//
+// Writers must not be used again after Close is called.
+func (w *Writer) Close() error {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	if w.batch == nil {
+		return nil
+	}
+	batch := w.batch
+	w.batch = nil
+	w.aggregator.writers.Delete(w)
```

Comment:
I think we can move managing the `writers` to the `aggregator` code? Also, maybe it would be beneficial to move the writer to `internal` package to make the interface (public and private methods) a bit more cleaner?

Reply:
> I think we can move managing the writers to the aggregator code?

Do you mean move move the `Writer.Close` method into aggregator?

> Also, maybe it would be beneficial to move the writer to internal package to make the interface (public and private methods) a bit more cleaner?

There's not a lot of code in `writer.go`, so I don't personally see value in making an internal package. I'm not sure I follow though.

Reply:
> There's not a lot of code in writer.go, so I don't personally see value in making an internal package. I'm not sure I follow though.

Sorry, should have explained in more detail. The interaction between writer and aggregator seems a bit cluttered since they use the private and the public methods as required. I was suggesting moving the `writer` to `internal` in its separate package so that the interaction is cleaner (the exposed public methods are called between writer and aggregator).

> Do you mean move move the Writer.Close method into aggregator?

I was referring to removing the access of `writers` map from writer.go. Maybe we can return the batch from `Close` and delete if from map in the `aggregator` code.

Reply:
> Sorry, should have explained in more detail. The interaction between writer and aggregator seems a bit cluttered since they use the private and the public methods as required. I was suggesting moving the writer to internal in its separate package so that the interaction is cleaner (the exposed public methods are called between writer and aggregator).

I had a look, but it's not straightforward since the writer needs to refer to the CombinedMetricsKey type.

> I was referring to removing the access of writers map from writer.go. Maybe we can return the batch from Close and delete if from map in the aggregator code.

That won't allow a direct call to `Writer.Close` to remove the writer from the Aggregator. A couple of options:
 - We can not remove from the map eagerly, but defer that to harvest or Aggregator.Close. I don't love this since it means resources hang around unnecessarily.
 - We could inject a callback into Writer, to be called on Close

Reply:
> We could inject a callback into Writer, to be called on Close

This sounds good but _it might_ be  problematic to pass callback in case we want to close the writer independently of the aggregator.